Feature/rabbit mq

the-man-w-laughs opened the pull request:
The message broker (RabbitMQ) is used to notify microservices about user creation or deletion. 
The Identity service acts as a publisher, while TeamHub and ReportHub act as consumers.

the-man-w-laughs committed:
Add message publisher to Identity.

Major changes:
- add UserCreatedMessage;
- add UserDeletedMessage.

the-man-w-laughs committed:
Add consumer to TeamHub.

Major changes:
- add create user handler;
- add delete user handler.

the-man-w-laughs committed:
Add consumer to ReportHub.

Major changes:
- add handler on users delete (all users projects reports will be deleted).

the-man-w-laughs committed:
Add repositories for Redis

Major changes:
- Add RedisRepository (base repository);
- Add UserRequestRepository (concrete implementation to store time of user's latest reuquest);

the-man-w-laughs committed:
Add cacher to ApiGateway;

Major changes:
- register IUserRequestRepository (to access redis);
- add UserChageMiddleware (to set last user's request).

the-man-w-laughs committed:
Add redis to TeamHub.

Major changes:
- add extra field to UserResponseDto (IsOnline);
- put additional logic to user automapper to make request to the redis to set isOnline field.

the-man-w-laughs committed:
Refactor ApiGateway

the-man-w-laughs committed:
Remove user caching from api gateway.

the-man-w-laughs committed:
Final implementation of redis caching

Major changes:
- add IsOnline property to UserDtos;
- add value resolvers for UserDtos in identity and teamhub (to set IsOnline property);
- add middleware to cache time of the latest user's request.

Description:
First I tried to put caching to api gateway but It didn't work at all, so I decided to put this middleware in each microservice.

mebr1 commented on src/services/Identity/Identity.WebAPI/Extensions/MessageBrokerExtensions.cs:
```diff
@@ -0,0 +1,43 @@
+                var host = config["MessageBroker:Host"];
+                var virtualHost = config["MessageBroker:VirtualHost"];
+                var username = config["MessageBroker:Username"];
+                var password = config["MessageBroker:Password"];
```
Let's use IOptions. This makes the code more secure and maintainable. Apply this to all methods with MassTransit configuration

mebr1 commented on src/services/Identity/Identity.WebAPI/Extensions/MessageBrokerExtensions.cs:
```diff
@@ -0,0 +1,43 @@
+                        cfg.Host(
+                            host,
+                            virtualHost,
+                            h =>
```
Let's use a more understandable name

mebr1 commented on src/services/TeamHub/TeamHub.BLL/MediatR/CQRS/User/QueryHandlers/GetAllUsersQueryHandler.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        var users = await _userRepository.GetAllAsync(cancellationToken);
+        var userResponseDtos = users.Select(project => _mapper.Map<UserResponseDto>(project));
+        return userResponseDtos;
```
New line before return

mebr1 commented on src/services/TeamHub/TeamHub.WebApi/controllers/UserController.cs:
```diff
@@ -0,0 +1,31 @@
+    /// Get Users
+    /// </summary>
+    [HttpGet]
+    public async Task<IActionResult> GetComment(CancellationToken cancellationToken)
```
For token you can use HttpContext.RequestAborted. In case of this you do not need to pass token in controller method. Only in handler

mebr1 commented on src/services/ReportHub/ReportHub.BLL/MassTransit/Consumers/UserDeletedMessageConsumer.cs:
```diff
@@ -0,0 +1,37 @@
+                project => project.ProjectAuthorId == context.Message.Id
+            );
+
+            foreach (var project in usersProjects)
+            {
+                foreach (var report in project.Reports)
+                {
+                    await _fileRepository.DeleteFileFromMinioAsync(report.Path);
+                }
+            }
```
You have a nested loop with database calls inside. It puts a significant load on a database

mebr1 commented on src/services/Identity/Identity.WebAPI/Extensions/MessageBrokerExtensions.cs:
```diff
@@ -0,0 +1,43 @@
+            ConfigurationManager config
+        )
+        {
+            services.AddMassTransit(x =>
```
Let's use a more understandable name

mebr1 commented on src/services/ReportHub/ReportHub.BLL/Extensions/MassTransitExtensions.cs:
```diff
@@ -0,0 +1,53 @@
+            IConfiguration config
+        )
+        {
+            services.AddMassTransit(x =>
```
Correct the naming in all services with MassTransit configuration. Make the naming more understandable

mebr1 commented on src/services/TeamHub/TeamHub.WebApi/controllers/UserController.cs:
```diff
@@ -0,0 +1,31 @@
+using Microsoft.AspNetCore.Mvc;
+using TeamHub.BLL.MediatR.CQRS.Users.Queries;
+
+namespace TeamHub.WebApi.controllers;
```
Your folder starts with a lowercase letter

the-man-w-laughs committed:
Refactor RabbitMQ configuration

Major changes:
- add RabbitMQSettings;
- rewrite MassTransit registration using IOptions;
- optimise multiple file deletion;
- make namespaces names start with capital letter.

the-man-w-laughs committed:
Merge pull request #9 from the-man-w-laughs/feature/redis

Feature/redis

the-man-w-laughs committed:
Make controller name plural.

the-man-w-laughs committed:
Refactor code

Major changes
- add async suffixes to all async methods.

the-man-w-laughs committed:
Merge branch 'feature/reports-microservice' into feature/rabbitMQ
